src/txn/txn_rollback_to_stable.c

```diff
@@ -358,6 +360,84 @@ __txn_abort_newer_col_fix(
         __txn_abort_newer_insert(session, ins, rollback_timestamp);
 }
 
+/*
+ * __txn_abort_row_hs_updates_using_build_in_mem_page --
+ *     Abort updates on a history store using the in-memory build reconciled page of data store.
+ */
+static int
+__txn_abort_row_hs_updates_using_build_in_mem_page(WT_SESSION_IMPL *session, const void *image,
+  const uint8_t *addr, size_t addr_size, wt_timestamp_t rollback_timestamp)
+{
+    WT_DECL_RET;
+    WT_ITEM tmp;
+    WT_PAGE *mod_page;
+    WT_ROW *rip;
+    uint32_t i, page_flags;
+    const void *image_local;
+
+    /*
+     * Don't pass an allocated buffer to the underlying block read function, force allocation of new
+     * memory of the appropriate size.
+     */
+    WT_CLEAR(tmp);
+
+    image_local = image;
+
+    if (image_local == NULL) {
+        WT_RET(__wt_bt_read(session, &tmp, addr, addr_size));
+        image_local = tmp.data;
+    }
+
+    page_flags = WT_DATA_IN_ITEM(&tmp) ? WT_PAGE_DISK_ALLOC : WT_PAGE_DISK_MAPPED;
+    WT_ERR(__wt_page_inmem(session, NULL, image_local, page_flags, &mod_page));
+    tmp.mem = NULL;
+    WT_ROW_FOREACH (mod_page, rip, i)
+        WT_ERR_NOTFOUND_OK(
+          __txn_abort_row_with_hs_value(session, mod_page, rip, rollback_timestamp, false));
+    __wt_page_out(session, &mod_page);
+
+    if (0) {
+err:
+        if (mod_page != NULL)
+            __wt_page_out(session, &mod_page);
```

Comment:
Why do this for the error path? Wouldn't eviction do this anyway?

Reply:
The page is freed in success path also. The page is not added into the ref as there is no corresponding ref is available until it gets evicted. As we have a hazard pointer, so eviction doesn't evict it.

Reply:
Oh, I see. Why do we only need to free `tmp` on error? Could this error handling code just have the `if (0) {` removed, and the two `__wt_page_out` calls merged to be just one?

Reply:
Fixed.